docs/source/guide/server/configuration.md
# Server configuration
Here is a list of the configuration parameters to use to customize a selfhosted codebox instance.

**Required parameters**
- `CODEBOX_EXTERNAL_URL`: the url of the codebox instance
- `CODEBOX_WILDCARD_DOMAIN`: codebox allows you to expose ports running HTTP-based services either public or with password authentication. The ports will be exposed through subdomains of this domain. You will need to define a DNS record with a name such as `*.codebox.my-domain.com`.

**Optional parameters (do not change them if you don't know what are you doing)**
- `CODEBOX_DEBUG`: `default: false` run server in debug mode, more logs about error will be printed, do not turn on this setting in production environment.
- `CODEBOX_DB_DRIVER`: `default: mysql` db driver to usen
- `CODEBOX_DB_HOST`: `default: localhost` the hostname/ip address where the dbms is running
- `CODEBOX_DB_PORT`: `default: 3306` the port where dbms is listening on
- `CODEBOX_DB_NAME`: `default: codebox` the name of the database
- `CODEBOX_DB_USER`: `default: codebox` the user used for authentication on dbms
- `CODEBOX_DB_PASSWORD`: `default: password` the password used for authentication on dbms
- `CODEBOX_SERVER_PORT`: `default: 8100` the port where codebox server is listening
- `CODEBOX_BG_TASKS_CONCURRENCY`: `default: 5` the concurrency of the background tasks
- `CODEBOX_REDIS_HOST`: `default: localhost` specifies the hostname or IP address of the Redis server.
- `CODEBOX_REDIS_PORT` `default: 6379` specifies the port number on which the Redis server is listening for connections
- `CODEBOX_DATA_PATH`: `default: ./data` the path where data is stored
- `CODEBOX_AUTH_COOKIE_NAME`: `default: codebox_auth_token` the name of the authentication cookie
- `CODEBOX_SUBDOMAIN_AUTH_COOKIE_NAME` `default: subdomain_codebox_auth_token` the name of the authentication cookie use in subdomains


CHANGELOG.md
# Changelog

All changes will be reported in this file

## [v0.0.51] - 2025-11-28

### Added

- Check version of runners

### Changed

- Bump runner to v0.0.24

## [v0.0.50] - 2025-11-24

### Fixed

- UI improvements
- Fixed name of downloaded CLI binaries

## [v0.0.49] - 2025-11-16

### Fixed

- Fixed nginx configuration
- Fixed ping of agents
- Fixed forward of http requests to agents

## [v0.0.48] - 2025-11-16

### Added 

- Security improvements
- Web terminal for workspace containers
- Support for custom git ssh command
- Support for macOS

## [v0.0.47] - 2025-08-30

### Added 

- Added react redux

### Fixed

- UI/UX improvements

### Updated

- Refactoring of admin interface

## [v0.0.46] - 2025-08-30

### Fixed

- Fixed a bug where the forward TCP API runner didn't load a relationship.

## [v0.0.45] - 2025-08-24

### Fixed

- Bug fix, devcontainer configuration in a subfolder

## [v0.0.44] - 2025-08-13

### Fixed

- bug fix

## [v0.0.43] - 2025-08-11

### Fixed

- bug fix

## [v0.0.42] - 2025-08-10

### Added

- added documentation for APIs

## [v0.0.41] - 2025-08-06

### Fixed

- bug fix

## [v0.0.40] - 2025-07-06

### Fixed

- bug fix

## [v0.0.39] - 2025-07-05

### Fixed

- fixed registration of a new runner

### Added

- added docs about runners

## [v0.0.38] - 2025-06-30

### Added

- Added cli command to change users password

## [v0.0.37] - 2025-06-28

### Fixed

- bug fix

## [v0.0.36] - 2025-06-25

### Fixed

- bug fix

## [v0.0.35] - 2025-06-15

### Fixed

- bug fix

## [v0.0.34] - 2025-06-14

### Changed

- show runner versions in admin area

# [v0.0.33] - 2025-06-14

### Fixed

- fixed bug on workspace deletion

## [v0.0.32] - 2025-06-08

### Changed

- moved all frontend http requests to a dedicated file

## [v0.0.31] - 2025-06-05

### Fixed

- bug fix

## [v0.0.30] - 2025-06-03

### Added

- added workspace templates

## [v0.0.22] - 2025-04-27

### Fixed

- bug fix

## [v0.0.21] - 2025-04-26

### Changed

- pretty error messages
- use cookie with another name for subdomains'authentication

## [v0.0.20] - 2025-04-26

### Fixed

- fixed authentication of subdomains

## [v0.0.19] - 2025-04-21

### Changed

- improved web interfaces
- improved authentication flow

## [v0.0.18] - 2025-04-17

### Added

- set git user.name and user.email inside containers
- expose some workspace related environment variables

## [v0.0.17] - 2025-04-14

### Changed

- Moved repository from github to gitlab

## [v0.0.16] - 2025-04-13

### Added

- Interfaces to manage exposed ports
- Runners management interfaces
- Users management interfaces

### Changed

- Runner connection using chisel

## [v0.0.15] - 2025-03-27

### Added

- Runner authentication


config/env.go
package config

import (
	"fmt"
	"os"

	"github.com/caarlos0/env"
	"github.com/joho/godotenv"
)

var ServerVersion = "dbg-9.9.9"

type EnvVars struct {
	ExternalUrl             string `env:"CODEBOX_EXTERNAL_URL,required"`
	WildcardDomain          string `env:"CODEBOX_WILDCARD_DOMAIN,required"`
	DebugEnabled            bool   `env:"CODEBOX_DEBUG" envDefault:"true"`
	DBDriver                string `env:"CODEBOX_DB_DRIVER" envDefault:"mysql"`
	DBHost                  string `env:"CODEBOX_DB_HOST" envDefault:"localhost"`
	DBPort                  int    `env:"CODEBOX_DB_PORT" envDefault:"3306"`
	DBName                  string `env:"CODEBOX_DB_NAME" envDefault:"codebox"`
	DBTestName              string `env:"CODEBOX_TEST_DB_NAME" envDefault:"codebox-test"`
	DBUser                  string `env:"CODEBOX_DB_USER" envDefault:"codebox"`
	DBPassword              string `env:"CODEBOX_DB_PASSWORD" envDefault:"password"`
	ServerPort              int    `env:"CODEBOX_SERVER_PORT" envDefault:"8080"`
	TasksConcurrency        int    `env:"CODEBOX_BG_TASKS_CONCURRENCY" envDefault:"5"`
	RedisHost               string `env:"CODEBOX_REDIS_HOST" envDefault:"localhost"`
	RedisPort               int    `env:"CODEBOX_REDIS_PORT" envDefault:"6379"`
	UploadsPath             string `env:"CODEBOX_DATA_PATH" envDefault:"./data"`
	UseSubDomains           bool   `env:"CODEBOX_USE_SUBDOMAINS" envDefault:"true"`
	AuthCookieName          string `env:"CODEBOX_AUTH_COOKIE_NAME" envDefault:"codebox_auth_token"`
	SubdomainAuthCookieName string `env:"CODEBOX_SUBDOMAIN_AUTH_COOKIE_NAME" envDefault:"subdomain_codebox_auth_token"`
	CliBinariesPath         string `env:"CODEBOX_CLI_BINARIES_PATH" envDefault:"./cli"`
	BaseDir                 string `env:"CODEBOX_BASE_DIR" envDefault:""`
}

var Environment *EnvVars

func InitCodeBoxEnv() error {
	codeboxEnvFilename := os.Getenv("CODEBOX_ENV_FILE")
	if codeboxEnvFilename == "" {
		codeboxEnvFilename = "codebox.env"
	}
	err := godotenv.Load(codeboxEnvFilename)
	if err != nil {
		return fmt.Errorf("failed to load environement variables from %s", codeboxEnvFilename)
	}

	e := EnvVars{}
	err = env.Parse(&e)
	if err != nil {
		return err
	}

	if err = os.MkdirAll(e.UploadsPath, 0777); err != nil {
		return err
	}

	Environment = &e
	return nil
}
